src/pybind/mgr/snap_schedule/fs/schedule.py

```diff
@@ -0,0 +1,311 @@
+"""
+from datetime import datetime, timezone
+import logging
+import sqlite3
+
+log = logging.getLogger(__name__)
+
+
+class Schedule(object):
+    '''
+    Wrapper to work with schedules stored in sqlite
+    '''
+    def __init__(self,
+                 path,
+                 schedule,
+                 retention_policy,
+                 fs_name,
+                 rel_path,
+                 start=None,
+                 subvol=None,
+                 created=None,
+                 first=None,
+                 last=None,
+                 last_pruned=None,
+                 created_count=0,
+                 pruned_count=0,
+                 active=True,
+                 ):
+        self.fs = fs_name
+        self.subvol = subvol
+        self.path = path
+        self.rel_path = rel_path
+        self.schedule = schedule
+        self.retention = retention_policy
+        if start is None:
+            now = datetime.now(timezone.utc)
+            self.start = datetime(now.year,
+                                  now.month,
+                                  now.day,
+                                  tzinfo=now.tzinfo)
+        else:
+            self.start = datetime.fromisoformat(start).astimezone(timezone.utc)
+        if created is None:
+            self.created = datetime.now(timezone.utc)
+        else:
+            self.created = datetime.fromisoformat(created)
+        if first:
+            self.first = datetime.fromisoformat(first)
+        else:
+            self.first = first
+        if last:
+            self.last = datetime.fromisoformat(last)
+        else:
+            self.last = last
+        if last_pruned:
+            self.last_pruned = datetime.fromisoformat(last_pruned)
+        else:
+            self.last_pruned = last_pruned
+        self.created_count = created_count
+        self.pruned_count = pruned_count
+        self.active = active
+
+    @classmethod
+    def _from_get_query(cls, table_row, fs):
+        return cls(table_row['path'],
+                   table_row['schedule'],
+                   table_row['retention'],
+                   fs,
+                   table_row['rel_path'],
+                   table_row['start'],
+                   table_row['subvol'],
+                   table_row['created'],
+                   table_row['first'],
+                   table_row['last'],
+                   table_row['last_pruned'],
+                   table_row['created_count'],
+                   table_row['pruned_count'],
+                   table_row['active'])
+
+    def __str__(self):
+        return f'''{self.path} {self.schedule} {self.retention}'''
+
+    CREATE_TABLES = '''CREATE TABLE schedules(
+        id integer PRIMARY KEY ASC,
+        path text NOT NULL UNIQUE,
+        subvol text,
+        rel_path text NOT NULL,
+        active int NOT NULL
+    );
+    CREATE TABLE schedules_meta(
+        id INTEGER PRIMARY KEY ASC,
+        schedule_id INT,
+        start TEXT NOT NULL,
```

Comment:
The reason I went with TEXT and iso datetime strings in UTC is that

1. sqlite doesn't know any date types. (see https://www.sqlite.org/datatype3.html)
2. No issues at all with with the whole 2038 problem.

The timestamp math that is used to calculate the next snapshot is the only part where unix epoch time makes sense (and that's what is used too). But otherwise it seems to be just as convenient to store timestamp strings.

Reply:
Right, but using `TIMESTAMP`, the raw type is actually TEXT. You can confirm:

````
$ python3
>>> import sqlite3
>>> import datetime
>>> db = sqlite3.connect('foo.db', detect_types=sqlite3.PARSE_DECLTYPES)
>>> c = db.execute('create table foo (bar integer, baz timestamp);')
>>> c=db.execute('insert into foo values(?, ?);', (2,datetime.datetime.now(),))
>>> c=db.execute('insert into foo values(?, ?);', (2,datetime.datetime.now(),))
>>> db.flush()
^D
$ sqlite3 foo.db
sqlite> select *, typeof(baz) from foo;
2|2020-05-19 09:07:16.595845|text
2|2020-05-19 09:08:45.850235|text
````

But, now that I look at this again. I see that the python3 adapter is inserting local times into the database. (`/headdesk`)

I was trying to be clever by suggesting the use of the python3 adapter which, now that I see it uses local time, really is utter shit. What I've done in the past for personal project is do:

`foo DATETIME NOT NULL DEFAULT (strftime('%s', 'now'))`

The `DATETIME` raw type is a integer for on-disk compactness. If you do basic dumps of the database it won't be readable as, obviously, it's a UNIX epoch. You can use strftime in your `SELECT `s to get a useful ISO string to pass to `datetime` (or just pass the UNIX epoch to datetime which is equiv.).

Anyway, this is all a suggestion.